### PR TITLE
Adding BWA-MEM

### LINKED ISSUE #953: Add BWA-MEM
Provide the option to use BWA-MEM for mapping of longer sequences and better performance

### PR DESCRIPTION
**Pull request by scarlhoff**
This closes #953 

## PR checklist

- [x] This comment contains a description of changes (with reason).
- [x] If you've fixed a bug or added code that should be tested, add tests!
    - [ ] If you've added a new tool - add to the software_versions process and a regex to `scrape_software_versions.py`
    - [x] If you've added a new tool - have you followed the pipeline conventions in the [contribution docs](<https://github.com/>nf-core/eager/tree/master/.github/CONTRIBUTING.md)
    - [ ] If necessary, also make a PR on the nf-core/eager _branch_ on the [nf-core/test-datasets](https://github.com/nf-core/test-datasets) repository.
- [x] Make sure your code lints (`nf-core lint .`).
- [x] Ensure the test suite passes (`nextflow run . -profile test,docker`).
- [x] Usage Documentation in `docs/usage.md` is updated.
- [x] Output Documentation in `docs/output.md` is updated.
- [ ] `CHANGELOG.md` is updated.
- [x] `README.md` is updated (including new tool citations and authors/contributors).

### CODE REVIEW

**Review by TCLamnidis — APPROVED** [review]
minor suggestions, but LGTM!

File: docs/output.md [hunk 1]
```diff
@@ -143,7 +143,7 @@ The resulting FASTQ files will only be present in your results directory if you
...
+  [BWA](https://bio-bwa.sourceforge.net/bwa.shtml) is a software package for mapping low-divergent sequences against a large reference genome, such as the human genome. It consists of three algorithms: BWA-backtrack (a.k.a `bwa aln`), BWA-SW and BWA-MEM (`bwa mem`). The first algorithm is designed for Illumina sequence reads up to 100bp, while the rest two for longer sequences ranged from 70bp to 1Mbp. BWA-MEM and BWA-SW share similar features such as long-read support and split alignment, but BWA-MEM, which is the latest, is generally recommended for high-quality queries as it is faster and more accurate. BWA-MEM also has better performance than BWA-backtrack for 70-100bp Illumina reads.
```
Comment (jfy133) [inline docs/output.md:146]: I wouldn't mention `bwa-sw`, we don't use it (and neither does anyone else I know)

File: subworkflows/local/map.nf [hunk 2]
```diff
@@ -3,9 +3,10 @@
 //
 
 include { FASTQ_ALIGN_BWAALN                             } from '../../subworkflows/nf-core/fastq_align_bwaaln/main'
+include { BWA_MEM                                        } from '../../modules/nf-core/bwa/mem/main'
 include { SAMTOOLS_MERGE                                 } from '../../modules/nf-core/samtools/merge/main'
 include { SAMTOOLS_SORT                                  } from '../../modules/nf-core/samtools/sort/main'
-include { SAMTOOLS_INDEX                                 } from '../../modules/nf-core/samtools/index/main'
+include { SAMTOOLS_INDEX as SAMTOOLS_INDEX_MEM; SAMTOOLS_INDEX as SAMTOOLS_INDEX_MERGE                                 } from '../../modules/nf-core/samtools/index/main'
 include { SAMTOOLS_FLAGSTAT as SAMTOOLS_FLAGSTAT_MAPPED  } from '../../modules/nf-core/samtools/flagstat/main'
```
Comment (jfy133) [inline subworkflows/local/map.nf:10]: Missing the Harshil Alignment™️ (@drpatelh )

File: nextflow_schema.json [hunk 3]
```diff
@@ -484,6 +484,18 @@
                     "default": 2,
                     "description": "Specify the -o parameter for BWA aln i.e. the number of gaps allowed.",
                     "help_text": "Configures the number of gaps used in bwa aln. Default is set to bwa default.\n\n> Modifies BWA aln parameter: `-o`"
+                },
+                "mapping_bwamem_k": {
+                    "type": "integer",
```
Comment (jfy133) [inline nextflow_schema.json:489]: Missing icon!

File: nextflow_schema.json [hunk 4]
```diff
@@ -484,6 +484,18 @@
...
+                    "default": 19,
+                    "description": "Specify the -k parameter for BWA mem i.e. the  minimum seed length",
+                    "help_text": "Configures the minimum seed length unsed in BWA-MEM. Default is set to BWA default.\n\n> Modifies BWA-MEM parameter: `-k`"
+                },
+                "mapping_bwamem_r": {
+                    "type": "number",
```
Comment (jfy133) [inline nextflow_schema.json:495]: Missing icon!

**Review by jfy133 — COMMENTED** [review]
some minor things but overall looking good!

File: conf/modules.config [hunk 5]
```diff
@@ -255,6 +252,20 @@ process {
...
     }
 
+    withName: BWA_MEM {
+        ext.args = { "-k ${params.mapping_bwamem_k} -r ${params.mapping_bwamem_r}" }
+        publishDir = [
+            enabled: false
+        ]
+    }
+
+    withName: ".*MAP:SAMTOOLS_INDEX_MEM" {
```
Comment (jfy133) [inline conf/modules.config:262]: I think we can even remove the ugly warning now as the name is unique
```suggestion
    withName: "SAMTOOLS_INDEX_MEM" {
```

File: conf/modules.config [hunk 6]
```diff
@@ -270,7 +281,7 @@ process {
         ]
     }
 
-    withName: ".*MAP:SAMTOOLS_INDEX" {
+    withName: ".*MAP:SAMTOOLS_INDEX_MERGE" {
```
Comment (jfy133) [inline conf/modules.config:284]:
```suggestion
    withName: "SAMTOOLS_INDEX_MERGE" {
```

**Review by jfy133 — APPROVED** [review]
Two last suggestions, and once test pass you can merge :rocket: thank you @scarlhoff !

**Pull request merged**